### PR TITLE
feat: add attributes for cloud-native deployment

### LINKED ISSUE #3168: Support capturing vpc details
### Area(s)

area:cloud, area:destination, area:network, area:server, area:source

### What's missing?

While working through https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/44697 it emerged that we don't have a way to generically capture the vpc which the traffic is coming from.

### Describe the solution you'd like

Introduce the below attributes to capture the vpc:
- *.vpc.name
- *.vpc.id

With the attributes residing within the 5 namespaces

## Additional context
Complements the introduction of the vlan sub namespace

### Tip

<sub>[React](https://github.blog/news-insights/product-news/add-reactions-to-pull-requests-issues-and-comments/) with 👍 to help prioritize this issue. Please use comments to provide useful context, avoiding `+1` or `me too`, to help us triage it. Learn more [here](https://opentelemetry.io/community/end-user/issue-participation/).</sub>

### PR DESCRIPTION
**Pull request by thisthat**
Fixes #3168

Old PR: https://github.com/open-telemetry/opentelemetry-specification/issues/3168 / https://github.com/open-telemetry/opentelemetry-specification/pull/3169

## Changes

This PR introduces semantic conventions for Cloud Native Deployments.

## Related issues 

- https://github.com/open-telemetry/opentelemetry-specification/issues/3168

FYI @dyladan @trask @ruflin @felixbarny @reyang @ChrsMark @jsuereth

### CODE REVIEW

File: specification/trace/semantic_conventions/deployment.md [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+Cloud-native applications are composed of one or several components that are usually called [workloads](https://kubernetes.io/docs/concepts/workloads/).
+Similarly to applications, each stage of the lifecycle of a workload deployment SHOULD be recorded as an Event on the span during
+which it occurred.
+
+<!-- semconv deployment.workload -->
+The event name MUST be `deployment.workload`.
+
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `deployment.workload.name` | string | The name of the single workload under deployment. | `podtato-head-hat`; `podtato-head-left-leg` | Required |
```
Comment (felixbarny) [inline specification/trace/semantic_conventions/deployment.md:68]: Sounds like this is essentially the same as `service.name`?

**Review by felixbarny — COMMENTED** [review]
Some of the concepts, such as application and workload aren't unique to k8s/cloud native deployments. I wouldn't segregate them under a specific `deployment` namespace but rather try to map them to more generic concepts that also make sense outside of cloud native deployments.

This is a good example where it would be good to have a common catalogue of attributes and logs and traces just reference from the common catalogue.

**Comment by reyang** [comment]
Related to https://github.com/open-telemetry/oteps/pull/223.

@horovits FYI.

File: specification/trace/semantic_conventions/deployment.md [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+| `deployment.workload.namespace` | string | The namespace in which the workload will be deployed. | `podtatohead` | Recommended |
```
Comment (jsuereth) [inline specification/trace/semantic_conventions/deployment.md:69]: `deployment.workload.{name|version|namespace}` seem to overlap with `service.*` attributes.   Should we just share them here?


Also `deployment.workload.status` seems to be the status of the deployment, less the workload itself.

File: specification/trace/semantic_conventions/deployment.md [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+Their execution SHOULD be recorded as an Event on the span during which it occurred.
+
+<!-- semconv deployment.task -->
+The event name MUST be `deployment.task`.
+
+| Attribute  | Type | Description  | Examples  | Requirement Level |
+|---|---|---|---|---|
+| `deployment.task.name` | string | The name that identifies the executed task. | `database-migration`; `report-metric`; `provision-infrastructure` | Required |
+| `deployment.task.status` | string | The status of the task execution. | `Pending` | Recommended |
+| `deployment.application.name` | string | The name of the application under deployment. | `podtato-head` | Required |
```
Comment (jsuereth) [inline specification/trace/semantic_conventions/deployment.md:97]: I don't understand the difference between workload + application here.  Can you outline that somewhere?

File: specification/trace/semantic_conventions/deployment.md (hunk 2 above)
Comment (thisthat) [reply specification/trace/semantic_conventions/deployment.md:69]: Thank @jsuereth for looking into this. I agree, these 3 attributes can be replaced with `service.*` attributes. 

> Also deployment.workload.status seems to be the status of the deployment, less the workload itself.

Deployment is an overloaded term, especially in K8s. Maybe `delivery` would be a better term 🤔 
If I read "status of the deployment", I think about the specific [Kubernetes Deployment CR](https://kubernetes.io/docs/concepts/workloads/controllers/deployment/), but here I'd like to capture also the different "deployment models" that K8s allows which are grouped around the term [Workload](https://kubernetes.io/docs/concepts/workloads/controllers/).

File: specification/trace/semantic_conventions/deployment.md (hunk 3 above)
Comment (thisthat) [reply specification/trace/semantic_conventions/deployment.md:97]: Sure, maybe this should be in the "Overview" section. 
The difference is that an application is an abstraction that groups together different workloads. For example, an application could be `my-awesome-website` which comprises 3 different workloads: frontend, backend, and database. 
Tasks are actions that are executed when delivering the application and its individual workloads. For example, a task could be sending a slack message that the delivery for a specific component started/finished. 
Do you think this explanation is clear @jsuereth? If so, I would add it to the document.

**Comment by thisthat** [comment]
> Does it make sense to spin up a CI/CD working group and have this as part of that?

I would be interested in such thing 👍

**Comment by joaopgrassi** [comment]
Hi @thisthat! Sorry for the ping, but is this something you are still interested in working on?

**Comment by thisthat** [comment]
Hey @joaopgrassi sure! I am still interested 👍

File: specification/trace/semantic_conventions/deployment.md [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+and are delivered mostly via either Operators or GitOps practices and tools.
+Hence, the delivery process is modeled as an event stream of OpenTelemetry [`Event`s](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/api.md#add-events).
+This document defines semantic conventions to collect telemetry signals about events that occur during the
+deployment of an application.
+
+## Definitions
+
+### Application
+
+Each stage of the lifecycle of a cloud-native application deployment SHOULD be recorded as an Event on the span during
```
Comment (christophe-kamphaus-jemmic) [inline specification/trace/semantic_conventions/deployment.md:34]: Should these events have any relation with the [Continuous Deployment Events](https://github.com/cdevents/spec/blob/v0.3.0/continuous-deployment.md) of [CDEvents](https://cdevents.dev/)?

**Comment by joaopgrassi** [comment]
Hi @thisthat !

We changed how the CHANGELOG.md is managed. Please take a look at https://github.com/open-telemetry/semantic-conventions/blob/main/CONTRIBUTING.md#adding-a-changelog-entry to see what needs to be done. Sorry for the disruption.

**Comment by jsuereth** [comment]
Let's close this PR as being stale.  We can reopen it with changes (including attribute registry usage) and align w/ exisitng CI/CD plans.

**Pull request closed without merging**